lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
+            return cls.branch(version_name)
+        # (4) Otherwise, we have ambiguous ref arguments, or no ref arguments at all,
+        # so error.
+        if num_specified != 1:
+            raise InvalidGitRef(dedent("""\
+            Exactly one of commit, tag, or branch must be specified for a git url.
+            Given:
+            commit={0}
+            tag={1}
+            branch={2}
+            """).format(commit, tag, branch))
+        if commit is not None:
+            return cls.commit(commit)
+        if tag is not None:
+            return cls.tag(tag)
+        assert branch is not None, (commit, tag, branch, version_name)
+        return cls.branch(branch)
+
+    def refspec(self):
+        # type: () -> str
+        """Return a string which unambiguously identifies this ref to git."""
+        if self.ref_type == 'commit':
+            return self._ref
+        if self.ref_type == 'tag':
+            return 'refs/tags/{0}'.format(self._ref)
+        assert self.ref_type == 'branch', self
+        return 'refs/heads/{0}'.format(self._ref)
+
+    def fetch_spec(self):
+        # type: () -> str
+        """Return an argument to update a local ref from the remote, if applicable."""
+        if self.ref_type == 'commit':
+            # Commits cannot be updated and do not require any special syntax to fetch.
+            return self.refspec()
+        # Otherwise, we use the syntax to update the local ref from a remote ref of
+        # the same name, fetching updates if available/applicable.
+        assert self.ref_type in ['tag', 'branch'], self
+        maybe_mutating_prefix = '+' if self.is_mutable() else ''
+        return '{0}{1}:{1}'.format(maybe_mutating_prefix, self.refspec())
+
+    def is_mutable(self):
+        # type: () -> bool
+        """Whether this type of git ref should check for updates to that ref."""
+        if self.ref_type == 'branch':
+            return True
+        assert self.ref_type in ['tag', 'commit'], self
+        return False
+
+    def repo_info_for_reference(self):
+        # type: () -> str
+        if self.ref_type == 'commit':
+            return 'at commit {0}'.format(self._ref)
+        if self.ref_type == 'tag':
+            return 'at tag {0}'.format(self._ref)
+        assert self.ref_type == 'branch', self
+        return 'on branch {0}'.format(self._ref)
+
+
+class GitFetchStageConfiguration(object):
+    """Validate parameters used to customize a specific git fetch operation.
+
+    This information is orthogonal to that which is processed in `GitRef`. It describes
+    what operations are performed to *prepare* a git checkout for a spack `Stage`,
+    *after* fetching the configured ref.
+    """
+    submodules = None           # type: bool
```

Comment:
These will always be shared among all classes instantiated at the same time? E.g., having them here vs. just set in init means that they will be.

Reply:
So yes, but [this is the way to tell mypy that a class has an instance variable with this type](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes). By doing this, we have a neat little description of all the instance variables that a class will have (and their types), without having to scan the body of `__init__` (which may change). This ideally makes any later diffs that modify the class variables much easier to review, because regardless of how `__init__` changes, it will be clear at a glance which variables are set.

I can't easily find the documentation about this, but I recall the best practice for py2-compatible code is to do `= None` for instance variables in order to make use of this paradigm. When we require py3 we can just type `submodules: bool`.